src/client/client_impl.cc

```diff
@@ -79,6 +82,189 @@ google::protobuf::util::Status ClientImpl::CreateFile(
+        chunk_server_service_client_[server_address]->SendRequest(
+            read_file_chunk_request, client_context));
+    
+    // Handle grpc error, log and continue to try the next chunk server
+    if (!read_file_chunk_reply_or.ok()) {
+      LOG(ERROR) << "Read file chunk " << chunk_handle << " from "
+                 << server_address << " failed due to " 
+                 << read_file_chunk_reply_or.status().error_message();
+      continue;
+    }
+
+    // Handle chunk status error, log and continue
+    auto read_file_chunk_reply(read_file_chunk_reply_or.ValueOrDie());
+    switch (read_file_chunk_reply.status()) {
+      case ReadFileChunkReply::UNKNOWN:
+        LOG(ERROR) << "Unknown error while reading " + chunk_handle;
+        continue; 
+      case ReadFileChunkReply::FAILED_NOT_FOUND:
+        LOG(ERROR) << "Chunk not found: " + chunk_handle;
+        continue;
+      case ReadFileChunkReply::FAILED_STALE_VERSION:
+        LOG(ERROR) << "Reading stale version " << std::to_string(chunk_version)
+                   << " for chunk " + chunk_handle;
+        continue;
+      default: break;
+    }
+
+    // This read has been successful, return here. We expect the compiler to
+    // execute copy elision here to "move" the read reply. (It is tempting 
+    // to explicitly write std::move(read_file_chunk_reply) but that is not 
+    // the right approach to do so). 
+    return read_file_chunk_reply;
+  }
+
+  // Failed to read from all chunk servers
+  return google::protobuf::util::Status(
+             google::protobuf::util::error::INTERNAL,
+             "Failed to read from all chunk servers for " + chunk_handle);
+}
+
+google::protobuf::util::StatusOr<std::pair<size_t, void*>> 
+    ClientImpl::ReadFile(const char* filename, size_t offset, size_t nbytes) {
+  // Record the number of bytes that we already read
+  size_t bytes_read(0);
+  // Record the number of bytes to read
+  size_t remain_bytes(nbytes);
+
+  // We simply allocate the necessary size for the buffer. There is a 
+  // possibility that we may not read up to nbytes, but it is ok to pass
+  // this optimization opportunity as it may cost more to realloc
+  void* buffer(malloc(nbytes));
+  if (!buffer) {
+    return google::protobuf::util::Status(
+               google::protobuf::util::error::RESOURCE_EXHAUSTED,
+               "Not enough memory: malloc fails");
+  }
+
+  size_t chunk_block_size(config_manager_->GetFileChunkBlockSize() * 
+                              common::bytes_per_mb);
+
+  for (size_t chunk_index = offset / chunk_block_size; 
+       chunk_index * common::bytes_per_mb < offset + nbytes;
+       chunk_index++) {
+    auto file_chunk_data_or(ReadFileChunk(filename, chunk_index, 
+                                          std::min(remain_bytes, 
+                                                   common::bytes_per_mb)));
+    // If one of the chunk's read fails, free the buffer and return 
+    if (!file_chunk_data_or.ok()) {
```

Comment:
This is not necessarily true. We can try to read past the EOD, so if the error is a file name with chunk index that's not found, it's either (1) an error if this is the starting block, (2) possible EOD so we do NOT fail but return the bytes read so far, (3) return an error for not being able to an intermediate chunk index (so if index 1 2 4 are ok but 3 is not, it's error)

Reply:
we prob should consider making all of these reads as threads/parallel, and not synchronous.

Reply:
I think it would be helpful if the ReadFileChunk could reply with a status as "EOF", that way the caller here knows clearly about reaching an EOF, instead of having to inferring this from the number of bytes read. The corner case here is that you do not know if you have reached an EOF if bytes_have_read == bytes_to_read, so you have to try the next chunk. 

I can leave a TODO here for this corner case.